### DISCUSSION TITLE
Lecture7-- testPriceFeedVersionIsAccurate test if getversion() is in PriceConverter library

### DISCUSSION THREAD

**Discussion by grmihadogr** [Cyfrin/foundry-full-course-cu#399] [Q&A] [answered]
In Lecture4, I followed the instruction to put function getversion() into PriceConverter.sol library. 

In FundMe.sol contract, I imported the PriceConverter and commented out original import. I have the following code 
```solidity
import {PriceConverter} from "./PriceConverter.sol";
//import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
```
In the PriceConverter library, I have the following code
```solidity
    function getversion() internal view returns (uint256) {
        uint256 theversion;
        return
            theversion = AggregatorV3Interface(0x694AA1769357215DE4FAC081bf1f309aDC325306).version();
    }
```

 Therefore, in my FundMe.sol contract, there is no function defined as getversion() and {AggregatorV3Interface} was not imported.

When I write the FundMeTest.t.sol for testPriceFeedVersionIsAccurate()
```solidity
 function testPriceFeedVersionIsAccurate() public {
        uint256 theversion = fundMe.getversion();
        assertEq(theversion, 4);
    }
```

I encountered following error 
```
Member "getversion" not found or not visible after argument-dependent lookup in contract FundMe.solidity(9582)
```

I think the reason is the function getversion() is an internal defined function in the library and not a public function defined in FundMe. 
If this is the reason, how can I fix it?

I also have looked at the code published on github, I found this code in lecture 4's FundMe.sol contract. It seems Patrick added the {AggregatorV3Interface} import and a public function getVersion() back into FundMe to avoid the problem I encountered. 
```solidity
import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";
```
```solidity
    function getVersion() public view returns (uint256){
        AggregatorV3Interface priceFeed = AggregatorV3Interface(0x694AA1769357215DE4FAC081bf1f309aDC325306);
        return priceFeed.version();
    }
```

**Comment by ghost** [comment, marked as answer]
Hello @grmihadogr, what you can do in the `FundMe` contract is the following:

```javascript
  function getPriceFeedVersion() public view returns (uint256) {
      return PriceConverter.getversion();
  }
```

And leave `getversion` in `PriceConverter` as internal. This should work.

**Reply by grmihadogr** [reply]
Hi @pacelliv 
Thank you so much for your reply and I tried that code on remix and it did return the correct version to me. 

I did more research and found this link 

https://solidity-by-example.org/library/

It seems there are 2 ways to interact with library. One is with keyword using and attach the functions to the data type after using. (this is also the way the lecture 4 introduced.)

The other way is directly usage. In that case, I should treat the library as a "instance of a contract"

**Comment by 0xKumarShivam** [comment]
Don't use your own file that you have created, just copy it from github repo.  it may help to avoid this issue.
